File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/viewmodels/ObjectivesViewModel.java
package com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels;

import androidx.lifecycle.ViewModel;

import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.missions.data.MissionsData;

/**
 * ObjectivesViewModel class
 *
 * @author TritiumGamingStudios
 */
public class ObjectivesViewModel extends ViewModel {

    private MissionsData.Objective[] objectivesSpinnerObjectives = null;
    private boolean[] isObjectiveCompleted = null;

    private String ghostName = null;

    private boolean responseState = false; // alone , group

    /*
     * Chosen spinner objectives
     */
    public void setObjectivesSpinnerObjectives(MissionsData.Objective[] objectiveSpinnerObjectives) {
        this.objectivesSpinnerObjectives = objectiveSpinnerObjectives;
    }

    public MissionsData.Objective[] getObjectivesSpinnerObjectives() {
        return objectivesSpinnerObjectives;
    }

    /*
     * Ghost name
     */
    public void setGhostName(String ghostName) {
        this.ghostName = ghostName;
    }

    public String getGhostName() {
        return ghostName;
    }

    /*
     * Response
     */
    public void setResponseState(boolean responseState) {
        this.responseState = responseState;
    }

    public boolean getResponseState() {
        return responseState;
    }

    /*
     * Objective Completed Buttons
     */
    public void setObjectiveCompletion(boolean[] isObjectiveCompleted) {
        this.isObjectiveCompleted = isObjectiveCompleted;
    }

    public boolean[] getObjectiveCompletion() {
        return isObjectiveCompleted;
    }

    /*
     * Reset
     */
    public void reset() {
        objectivesSpinnerObjectives = null;
        isObjectiveCompleted = null;
        ghostName = null;
        responseState = false;
    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/viewmodels/ObjectivesViewModel.kt
package com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels

import android.content.Context
import androidx.lifecycle.ViewModel
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.models.missions.MissionsListModel

class ObjectivesViewModel : ViewModel() {

    companion object {
        const val UNKNOWN: Response = 0
        const val ALONE: Response = 1
        const val GROUP: Response = 2

        const val NOT_COMPLETE: MissionStatus = false
        const val COMPLETE: MissionStatus = true
    }

    /* All possible objectives */
    var missionsListModel: MissionsListModel? = null
    /* Objective Completed Buttons */
    var spinnerCompletionStatus: BooleanArray = BooleanArray(3) { NOT_COMPLETE }

    /* Ghost name */
    var ghostName: String? = null
        get() { return (field ?: "")}

    /* Response */
    var responseState: Response = UNKNOWN // alone , group

    fun init(context: Context) {
        missionsListModel = missionsListModel ?: MissionsListModel(context)
    }

    fun toggleCompletionStatus(spinnerIndex: Int) {
        spinnerCompletionStatus[spinnerIndex] = !spinnerCompletionStatus[spinnerIndex]
    }

    fun reset() {
        ghostName = null
        missionsListModel?.reset()
        spinnerCompletionStatus.all { NOT_COMPLETE }
        responseState = UNKNOWN
    }

}

typealias Response = Int
typealias MissionStatus = Boolean
